### PR TITLE
[SPARK-25811][PySpark] Raise a proper error when unsafe cast is detected by PyArrow

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Since 0.11.0, PyArrow supports to raise an error for unsafe cast ([PR](https://github.com/apache/arrow/pull/2504)). We should use it to raise a proper error for pandas udf users when such cast is detected.

Added a SQL config `spark.sql.execution.pandas.arrowSafeTypeConversion` to disable Arrow safe type check.

## How was this patch tested?

Added test and manually test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @BryanCutler

File: python/pyspark/serializers.py [hunk 1]
```diff
@@ -248,7 +248,14 @@ def create_array(s, t):
             # TODO: see ARROW-2432. Remove when the minimum PyArrow version becomes 0.10.0.
             return pa.Array.from_pandas(s.apply(
                 lambda v: decimal.Decimal('NaN') if v is None else v), mask=mask, type=t)
-        return pa.Array.from_pandas(s, mask=mask, type=t)
+        try:
+            array = pa.Array.from_pandas(s, mask=mask, type=t)
+        except pa.ArrowInvalid as e:
```
Comment (BryanCutler) [inline python/pyspark/serializers.py]: I'm not sure if there are other issues that could case an `ArrowInvalid` error. What do you think about catching an `ArrowException` and making the message a little more generic?

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -4961,6 +4961,31 @@ def foofoo(x, y):
...
+        df = self.spark.createDataFrame(pdf).repartition(1)
+
+        @pandas_udf(returnType="int")
+        def udf(column):
+            return pd.Series(np.linspace(0, 1, 3))
+
+        udf_boolean = df.select(['A']).withColumn('udf', udf('A'))
+
+        # Since 0.11.0, PyArrow supports the feature to raise an error for unsafe cast.
+        if LooseVersion(pa.__version__) >= LooseVersion("0.11.0"):
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: I checked how 0.8.0 is working and it does raise an error for something like overflows, but not truncation like this test. Can you also add a check for overflow not dependent on the version?

**Review by BryanCutler — COMMENTED** [review]
Thanks @viirya , looks good so far! Were you thinking of putting in a spark config to toggle the `safe` flag here or in a follow up? I think we really need that because prior to v0.11.0 it allowed unsafe casts, but now raises an error by default.

File: python/pyspark/serializers.py (hunk 1 above)
Comment (viirya) [reply python/pyspark/serializers.py]: I agree that is better. I will update this later.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/tests.py]: Ok, sure.

**Comment by viirya** [comment]
@BryanCutler Thanks for looking at this!

Yea, this is a WIP work for early review and I will add a config to toggle the `safe` flag later.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests.py]: BTW, let's bump up the minimal required PyArrow and Pandas version up if possible at 3.0 :-)

File: python/pyspark/serializers.py [hunk 3]
```diff
@@ -284,7 +284,14 @@ def create_array(s, t):
         elif LooseVersion(pa.__version__) < LooseVersion("0.11.0"):
             # TODO: see ARROW-1949. Remove when the minimum PyArrow version becomes 0.11.0.
             return pa.Array.from_pandas(s, mask=mask, type=t)
-        return pa.Array.from_pandas(s, mask=mask, type=t, safe=False)
+        try:
+            array = pa.Array.from_pandas(s, mask=mask, type=t, safe=True)
+        except pa.ArrowException as e:
```
Comment (viirya) [inline python/pyspark/serializers.py]: @BryanCutler Now it catches `ArrowException`.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/tests.py]: @BryanCutler Do you mean this same udf raises an error like overflows? I tried with 0.8.0 but it doesn't? Am I missing something?

**Comment by viirya** [comment]
Sorry for not pushing this for a while. Hope we can make it soon.

**Comment by viirya** [comment]
> Were you thinking of putting in a spark config to toggle the safe flag here or in a follow up? I think we really need that because prior to v0.11.0 it allowed unsafe casts, but now raises an error by default.

Do we have proper way to get spark config value at executor side like serializers here? I found that in `read_udfs` some configs for pandas_udf evaluation are read from stream, is it the only way we have?

**Comment by ueshin** [comment]
So since pyarrow 0.11, we won't allow nullable integral types by default?
`ScalarPandasUDFTests.test_vectorized_udf_null_(byte|short|int|long)` will fail as per #23305, IIUC.

**Comment by ueshin** [comment]
@viirya We can use `runner_conf` as we use to provide the session timezone to `ArrowStreamPandasSerializer`. See `worker.py` and `ArrowUtils.getPythonRunnerConfMap()`.
Also, although I've never tried yet, `TaskContext` might have the configs in its local properties.

**Comment by viirya** [comment]
@ueshin Thanks. I've noticed that way, but wondering if it is only way to do that. I will use it to set safe check.

File: python/pyspark/worker.py [hunk 4]
```diff
@@ -253,7 +253,7 @@ def read_udfs(pickleSer, infile, eval_type):
 
         # NOTE: if timezone is set here, that implies respectSessionTimeZone is True
         timezone = runner_conf.get("spark.sql.session.timeZone", None)
-        ser = ArrowStreamPandasSerializer(timezone)
+        ser = ArrowStreamPandasSerializer(timezone, runner_conf)
```
Comment (BryanCutler) [inline python/pyspark/worker.py]: I think it's slightly better to parse the `runner_conf` for the needed options here and pass in the required flags like we do with `timezone`, instead of passing around the `runner_conf`

File: python/pyspark/sql/session.py [hunk 5]
```diff
@@ -556,8 +556,10 @@ def _create_from_pandas_with_arrow(self, pdf, schema, timezone):
         pdf_slices = (pdf[start:start + step] for start in xrange(0, len(pdf), step))
 
         # Create Arrow record batches
+        runner_conf = {"spark.sql.execution.pandas.arrowSafeTypeConversion":
+                       self._wrapped._conf.arrowSafeTypeConversion()}
         batches = [_create_batch([(c, t) for (_, c), t in zip(pdf_slice.iteritems(), arrow_types)],
-                                 timezone)
+                                 timezone, runner_conf)
```
Comment (BryanCutler) [inline python/pyspark/sql/session.py]: here, it doesn't really make sense to use a `runner_conf`, so it would be better to just pass in the flag

File: python/pyspark/serializers.py [hunk 6]
```diff
@@ -284,7 +284,19 @@ def create_array(s, t):
         elif LooseVersion(pa.__version__) < LooseVersion("0.11.0"):
             # TODO: see ARROW-1949. Remove when the minimum PyArrow version becomes 0.11.0.
             return pa.Array.from_pandas(s, mask=mask, type=t)
-        return pa.Array.from_pandas(s, mask=mask, type=t, safe=False)
+
+        enabledArrowSafeTypeCheck = \
+            runner_conf.get("spark.sql.execution.pandas.arrowSafeTypeConversion", "true") == 'true'
```
Comment (BryanCutler) [inline python/pyspark/serializers.py]: might want to tack on `.lower()` to ensure you are checking lower case

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 7]
```diff
@@ -138,36 +138,44 @@ def test_vectorized_udf_null_boolean(self):
...
 
     def test_vectorized_udf_null_int(self):
-        data = [(None,), (2,), (3,), (4,)]
-        schema = StructType().add("int", IntegerType())
-        df = self.spark.createDataFrame(data, schema)
-        int_f = pandas_udf(lambda x: x, IntegerType())
-        res = df.select(int_f(col('int')))
-        self.assertEquals(df.collect(), res.collect())
+        with self.sql_conf({
+                "spark.sql.execution.pandas.arrowSafeTypeConversion": False}):
```
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py]: This and other tests fail if `arrowSafeTypeConversion=True`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -1331,6 +1331,16 @@ object SQLConf {
...
 
+  val PANDAS_ARROW_SAFE_TYPE_CONVERSION =
+    buildConf("spark.sql.execution.pandas.arrowSafeTypeConversion")
+      .internal()
+      .doc("When true, enabling Arrow do safe type conversion check when converting" +
+        "Pandas.Series to Arrow Array during serialization. Arrow will raise errors " +
+        "when detecting unsafe type conversion. When false, disabling Arrow's type " +
+        "check and do type conversions anyway.")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: If enabling by default causes a lot of the current tests to fail, it might also do the same for users - maybe we should disable by default?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -1331,6 +1331,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val PANDAS_ARROW_SAFE_TYPE_CONVERSION =
+    buildConf("spark.sql.execution.pandas.arrowSafeTypeConversion")
+      .internal()
+      .doc("When true, enabling Arrow do safe type conversion check when converting" +
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: maybe reword to "When true, Arrow will perform safe type conversion when converting " +

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -1331,6 +1331,16 @@ object SQLConf {
...
+        "Pandas.Series to Arrow Array during serialization. Arrow will raise errors " +
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit: don't capitalize Array

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 11]
```diff
@@ -1331,6 +1331,16 @@ object SQLConf {
...
+        "when detecting unsafe type conversion. When false, disabling Arrow's type " +
```
Comment (BryanCutler) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: it might be good to elaborate what is an unsafe conversion, e.g. overflow

**Review by BryanCutler — COMMENTED** [review]
Thanks for getting back to this @viirya! Looks pretty good, I just think maybe parse the `runner_conf` earlier and use a `safe` flag in the ArrowPandasSerializer`.  Also, I'm a little concerned about the default value that might break a lot of people's code... but also if it is false by default, then it will allow overflows when it didn't before, so I'm not sure what's best.

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py (hunk 7 above)
Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py]: Yes, please see @ueshin's comment https://github.com/apache/spark/pull/22807#issuecomment-452193929.

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/tests.py]: Thanks @BryanCutler. It is clear to me now. I added a test for this.

File: python/pyspark/worker.py (hunk 4 above)
Comment (viirya) [reply python/pyspark/worker.py]: Ok. I passed in the required flag now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: As you said, if it is false by default, it allows overflow. cc @ueshin @HyukjinKwon  What do you think? Which one is better default value?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 11 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Added.

**Comment by viirya** [comment]
Thanks @BryanCutler for continuing to review this. I addressed above comments. We need to decide the default config value.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 10 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Fixed.

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py (hunk 7 above)
Comment (BryanCutler) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py]: I see, it's because of the NULL values

File: python/pyspark/sql/tests.py (hunk 2 above)
Comment (ueshin) [reply python/pyspark/sql/tests.py]: So pyarrow 0.8 - 0.10 raises an error for an integer overflow but not for cast from float to integral types, whereas 0.11 raises an error for both?

Comment (ueshin) [reply python/pyspark/sql/tests.py]: Seems like the cast from float to integral types are always working without an error in pyarrow < 0.11.

```python
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8())
<pyarrow.lib.Int8Array object at 0x11c42d940>
[
  -128
]
```

Comment (viirya) [reply python/pyspark/sql/tests.py]: Yeah but in pyarrow 0.11.0+ you'd see an error:
```python
>>> import pandas as pd
>>> import pyarrow as pa
>>> pa.__version__
'0.11.1'
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 474, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 169, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 69, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 81, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Floating point value truncated
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8(), safe=False)
<pyarrow.lib.Int8Array object at 0x7f3ee1a4b868>
[
  -128
]
```

Comment (ueshin) [reply python/pyspark/sql/tests.py]: Yeah. I was thinking how we should handle this behavior change.
We will have the behavior change anyway regardless of the config, right?
- `safe=True`: we can't use nullable integral types
- `safe=False`: we can't detect an integer overflow

Comment (viirya) [reply python/pyspark/sql/tests.py]: I think so. `safe=False` does the type conversion anyway even an overflow.

Comment (viirya) [reply python/pyspark/sql/tests.py]: If using `safe=False` by default, it is possible to break user's code without notifications. If using `safe=True` by default, it still breaks user's code, but there is error message so it should let user know what's happen.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I'd favor true by default. Do we even need this flag? As in many such cases, I'm just not clear a) how a user would find this option and b) when they would disable it. Disabling allows something to continue that is going to also fail or give an incorrect answer, right?

Comment (BryanCutler) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think the big issue with this is when NULL values are introduced in an integer column. Pandas will automatically convert these to floating-points to represent the NULLs, then when Arrow casts it back to integer, it will raise an error due to truncation - I don't think Arrow checks the actual values, but maybe it should?  For example, with safe=True:

```python
>>> pa.Array.from_pandas(pd.Series([1, None]), type=pa.int32(), safe=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 474, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 169, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 69, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 81, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Floating point value truncated
```

Comment (BryanCutler) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Does Spark check for these types of errors with standard udfs, when not using Arrow?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think for standard udfs, there is no similar type check:

```python
>>> test_udf = udf(lambda x: 128, "byte")
>>> spark.range(0, 1).select(test_udf('id')).show()
+------------+
|<lambda>(id)|
+------------+
|        -128|
+------------+
```

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: For arrow users, the old behavior is to have some errors when overflows happen. I think this is what users expect. If we make this config false by default, we change this behavior. This sounds not pretty safe.

Although we change behavior too if we make the config true by default. The behavior is changed by raising errors in cases including NULL values in an integer column. Because we show some messages and users can disable this config if they want.

We will change behavior either using true or false as default config. Sounds like making it as true is a bit safer.

Comment (kszucs) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: https://issues.apache.org/jira/browse/ARROW-4258

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks @kszucs

**Comment by srowen** [comment]
So, in these situations, we agree some error should occur. The current one is an overflow error -- right? or does it silently overfloat? Non-arrow UDFs seem to do the latter.

Based on that, I'm not clear the Arrow behavior should be different, especially if it already matches that behavior, and is what Arrow users expect.

No, this isn't a case for a flag. That is a failure to decide, punted to the user.

**Comment by BryanCutler** [comment]
@srowen we discussed adding this flag a while ago when pyarrow 0.11.1 made this option which seemed to change default behavior in pyspark.  Our intention was to preserve default behavior in pyspark and provide the user a config to change it if needed. It was only recently we learned there was a behavior change either way the config is set.

It does seem like there is a bug in Arrow that is affecting this too, https://issues.apache.org/jira/browse/ARROW-4258, although we still should make a patch for v0.11.1

**Comment by srowen** [comment]
I see; is it still reasonable to preserve the Pyspark behavior here? does the user really need to select a different type of failure mode? maybe, just based on my limited understanding, not clear.

**Comment by BryanCutler** [comment]
Here is the behaviour of PySpark for integer overflow:

_ | Integer Overflow
---|---
**PySpark Non-Arrow** | silent overflow
**with pyarrow version < 0.11.1**  | silent overflow 
**version >= 0.11.1 safe=False**  | silent overflow 
**version >= 0.11.1 safe=True** | raise "ArrowInvalid: Integer value out of bounds"

**Comment by srowen** [comment]
Can users directly set 'safe' if they want to? is it an Arrow option that users can access in Pyspark? if so then do we need to thunk it through in a Pyspark option? If not, I get it, we have to provide some way to pipe through that setting.

**Comment by BryanCutler** [comment]
>Can users directly set 'safe' if they want to?

No, there is no way for a user to set it. It's an option in pyarrow that is done when converting the returned Pandas Series back into Arrow data.

**Comment by srowen** [comment]
Got it, makes sense to me, with a default config value that sets safe=False.

**Comment by viirya** [comment]
@BryanCutler Thanks for making the summary. But is it correct that with pyarrow version < 0.11.1 "Integer Overflow" gets silent overflow? I recall that it throws an error: https://github.com/apache/spark/pull/22807#discussion_r246215883

**Comment by BryanCutler** [comment]
> But is it correct that with pyarrow version < 0.11.1 "Integer Overflow" gets silent overflow? I recall that it throws an error: #22807 (comment)

Uggh, yeah sorry, let me fix that

**Comment by viirya** [comment]
@BryanCutler @srowen Based on the summary, to set a false for the config value, I think we should also leave some words in migration guide for the behavior change. What do you think?

**Comment by srowen** [comment]
To be clear, if the new setting is false, then that enables the 'silent' behavior? that seems like the right default. Yes, can't hurt to note this in release notes ('release-notes' label in JIRA and 'Docs text')

**Comment by BryanCutler** [comment]
I think a default setting of false and a note in the migration guide sound like the best option.

**Comment by viirya** [comment]
Changed default config value to false and added a note to migration guide. Please take a look when you have time. Thanks.

**Review by srowen — APPROVED** [review]
Looks good pending tests

File: python/pyspark/sql/tests/test_pandas_udf.py [hunk 12]
```diff
@@ -197,6 +197,66 @@ def foofoo(x, y):
             ).collect
         )
 
+    def test_pandas_udf_detect_unsafe_type_conversion(self):
+        from distutils.version import LooseVersion
+        from pyspark.sql.functions import pandas_udf
```
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_pandas_udf.py]: nit: after the unit test reorg, the `pandas_udf` import is at the top so it's not needed here or in the other test

**Review by BryanCutler — APPROVED** [review]
LGTM  except for just one minor nit

File: docs/sql-migration-guide-upgrade.md [hunk 13]
```diff
@@ -41,6 +41,54 @@ displayTitle: Spark SQL Upgrading Guide
...
+          </th>
+          <th>
+            <b>Raise error</b>
+          </th>
+          <th>
+            <b>Raise error</b>
+          </th>
+        </tr>
+  </table>
+
```
Comment (BryanCutler) [inline docs/sql-migration-guide-upgrade.md]: Looks good, thanks for adding this!

File: python/pyspark/sql/tests/test_pandas_udf.py (hunk 12 above)
Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf.py]: Removed it. Thanks!

File: python/pyspark/sql/tests/test_pandas_udf.py [hunk 14]
```diff
@@ -197,6 +197,64 @@ def foofoo(x, y):
...
+            with self.sql_conf({
+                    "spark.sql.execution.pandas.arrowSafeTypeConversion": True}):
+                with self.assertRaisesRegexp(Exception,
+                                             "Exception thrown when converting pandas.Series"):
+                    df.select(['A']).withColumn('udf', udf('A')).collect()
+
+        # Disabling Arrow safe type check.
+        with self.sql_conf({
+                "spark.sql.execution.pandas.arrowSafeTypeConversion": False}):
+            df.select(['A']).withColumn('udf', udf('A')).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf.py]: nit: Can we asset the result too?

File: docs/sql-migration-guide-upgrade.md [hunk 15]
```diff
@@ -41,6 +41,54 @@ displayTitle: Spark SQL Upgrading Guide
...
+          <th>
+            <b>Silent overflow</b>
+          </th>
+          <th>
+            <b>Silently allows</b>
+          </th>
+        </tr>
+        <tr>
+          <th>
+            <b>version > 0.11.0, arrowSafeTypeConversion=true</b>
```
[the remark this answers is not in the document]
Comment (viirya) [reply docs/sql-migration-guide-upgrade.md]: I think this config can be kept even we set the minimal Arrow version to 0.12.0. If anything goes wrong, users still can disable the type check.

For now, there isn't consistent behavior across integer overflow and float point truncation. Either `true` or `false` causes behavior change. It is `false` by default to make it consistent to non-arrow UDFs.

If we are going to change it to `true` in the future, isn't a behavior change again?

File: python/pyspark/sql/tests/test_pandas_udf.py [hunk 16]
```diff
@@ -197,6 +197,64 @@ def foofoo(x, y):
...
 
+    def test_pandas_udf_detect_unsafe_type_conversion(self):
+        from distutils.version import LooseVersion
+        import pandas as pd
+        import numpy as np
+        import pyarrow as pa
+
+        values = [1.0] * 3
+        pdf = pd.DataFrame({'A': values})
+        df = self.spark.createDataFrame(pdf).repartition(1)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf.py]: sorry if I missed something. Why should we repartition?

File: python/pyspark/sql/tests/test_pandas_udf.py [hunk 17]
```diff
@@ -197,6 +197,64 @@ def foofoo(x, y):
...
+        with self.sql_conf({
+                "spark.sql.execution.pandas.arrowSafeTypeConversion": False}):
+            df.select(['A']).withColumn('udf', udf('A')).collect()
+
+    def test_pandas_udf_arrow_overflow(self):
+        from distutils.version import LooseVersion
+        import pandas as pd
+        import pyarrow as pa
+
+        df = self.spark.range(0, 1)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf.py]: nit .. : `spark.range(0)`.

File: python/pyspark/sql/tests/test_pandas_udf.py [hunk 18]
```diff
@@ -197,6 +197,64 @@ def foofoo(x, y):
...
+
+            # Disabling safe type check, let Arrow do the cast anyway.
+            with self.sql_conf({"spark.sql.execution.pandas.arrowSafeTypeConversion": False}):
+                df.withColumn('udf', udf('id')).collect()
+        else:
+            # SQL config `arrowSafeTypeConversion` no matters for older Arrow.
+            # Overflow cast causes an error.
+            with self.sql_conf({"spark.sql.execution.pandas.arrowSafeTypeConversion": False}):
+                with self.assertRaisesRegexp(Exception,
+                                             "Integer value out of bounds"):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf.py]: looks it can be inlined

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: No .. I was referring https://github.com/apache/spark/pull/22807#discussion_r247487281 ..

File: docs/sql-migration-guide-upgrade.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply docs/sql-migration-guide-upgrade.md]: Hmhmm .. yea .. Good point about consistency with the regular UDFs. I was thinking of 1. 
 removing such configurations out eventually (personally I don't like the bunch of configurations we have currently ..), 2. making all UDFs to expect the exact types (<- not sure yet it needs discussion). Yes, we can talk later ..

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Eh, it's okie. minor is minor. let's fix it later when we touch that code.

**Comment by viirya** [comment]
@HyukjinKwon ok. no problem.

File: python/pyspark/sql/tests/test_pandas_udf.py (hunk 16 above)
Comment (viirya) [reply python/pyspark/sql/tests/test_pandas_udf.py]: I thought I was writing this to make sure all values are in single partition so it matches the length of returned Pandas Series. This can be simplified. We can do this if we touch the code here next time.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (BryanCutler) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: > So, NULL -> Integer issue will be fixed in Arrow 0.12.0?

Yes, I verified this and ran the unit tests on the release

File: docs/sql-migration-guide-upgrade.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (BryanCutler) [reply docs/sql-migration-guide-upgrade.md]: Yeah, I think the config should stay too since it is only able to be set through pyarrow

**Comment by BryanCutler** [comment]
Thanks @viirya !
